Add support for checking if an `apt::keyring` is up-to-date with checksums

NeatNerdPrime opened the pull request:
https://github.com/puppetlabs/puppetlabs-apt/issues/1196

* Adds checksum and checksum_value parameter to apt::keyring, this should address issue/1196 as commented here https://github.com/puppetlabs/puppetlabs-apt/issues/1196#issuecomment-2329987738
* Includes tests, all green.

## Summary
Provide a detailed description of all the changes present in this pull request.

## Additional Context
Add any additional context about the problem here. 
- [ ] Root cause and the steps to reproduce. (If applicable)
- [ ] Thought process behind the implementation.

## Related Issues (if any)
Mention any related issues or pull requests.

## Checklist
- [ ] 🟢 Spec tests.
- [ ] 🟢 Acceptance tests.
- [ ] Manually verified. (For example `puppet apply`)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/puppetlabs/puppetlabs-apt?pullRequest=1199) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/puppetlabs/puppetlabs-apt?pullRequest=1199) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/puppetlabs/puppetlabs-apt?pullRequest=1199) it.</sub>

smortex reviewed (changes requested):
That looks good but we also need to adjust the `apt::source` defined type so that we can pass the `checksum` / `checksum_value` there too through the `key` parameter.

NeatNerdPrime commented:
Hello again.

I added the requested changes, but whiel doing so i stumbled upon a perhaps uncomfortable situation.

Given the changes in the source_spec.rb test: https://github.com/puppetlabs/puppetlabs-apt/pull/1199/files#diff-abe1d0bffeeb2d6d33970b8d3272b76455f72cf2541de5f85678bd76817446e4

I realize that the parameter 'key' in `apt::source` is misleading, since it actually refers to the (deprecated) method of managing keys with the `apt::key` defined resource. But at the same time it is also used for `apt::keyring`? That just don't feel alright to me. 

It begs the question: "Should we add a parameter 'keyring' to the defined resource, being mutually exclusive to the 'key' parameter, that calls upon `apt::keyring` and `apt::key` respectively?" 

This could make things more clear from a user perspective and  - unless Im mistaken - should be backwards compatible.

What do you people think?

skyamgarp commented:
@NeatNerdPrime , could you please update the PR?

NeatNerdPrime committed:
(issues/1196) Add checksum support to apt::keyring

Add checksum and checksum_value parameters to apt::keyring defined type,
enabling integrity verification of keyring files from remote HTTP sources
that do not provide the necessary headers for the file resource to detect
content changes.

Supported checksum types: md5, sha256, sha224, sha384, sha512.

Ref: https://github.com/puppetlabs/puppetlabs-apt/issues/1196

NeatNerdPrime committed:
(issues/1196) Pass checksum params through apt::source to apt::keyring

Trickle up the checksum-related parameters from apt::keyring to
apt::source, allowing users to specify checksum and checksum_value
in the key hash when declaring an apt::source resource.

Update documentation examples, key param description, REFERENCE.md,
and source spec tests to cover the new parameters.

Ref: https://github.com/puppetlabs/puppetlabs-apt/issues/1196

NeatNerdPrime committed:
Fix pre-existing documentation warnings and align parameter defaults

- Remove blank line before `define apt::auth` that caused puppet-strings
  to miss the @summary tag, resulting in an undocumented defined type.
- Shorten apt_key @summary to under 140 characters and move the detail
  into the description body.
- Align parameter default `=` signs across all manifests for consistent
  readability: auth, backports, conf, init, key, pin, ppa, setting,
  and source.
- Regenerate REFERENCE.md.

Documentation coverage is now 100%.

NeatNerdPrime commented:
@skyamgarp done, please check

NeatNerdPrime committed:
Merge branch 'main' into issues/1196

NeatNerdPrime committed:
Merge branch 'main' into issues/1196

skyamgarp commented:
@NeatNerdPrime 

Review: Request Changes

1. This PR **removes 7 existing test contexts** from `spec/defines/keyring_spec.rb`. The new checksum tests should be **added alongside** the existing ones, not substituted for them.
2. Same for `source_spec.rb`, do not alter existing tests unless they are failing and add new tests to test new changes.
3. Regression: `require => $require_dir` Removed
In `manifests/keyring.pp`, the `require => $require_dir` attribute has been silently dropped from the `file` resource block.

Please restore:
```
```puppet
file { $file:
  ensure         => 'file',
  mode           => $mode,
  owner          => 'root',
  group          => 'root',
  source         => $source,
  content        => $content,
  checksum       => $checksum,
  checksum_value => $checksum_value,
  require        => $require_dir,  # ← this must be restored
}

```
4. Missing Parameter Pair Validation

There is no guard to catch the case where only one of `checksum` / `checksum_value` is provided. Passing one without the other will produce a confusing runtime error from Puppet's `file` resource rather than a clear module-level message. Please add validation:

```puppet
if $checksum and !$checksum_value {
  fail("apt::keyring: 'checksum_value' is required when 'checksum' is specified")
}
if $checksum_value and !$checksum {
  fail("apt::keyring: 'checksum' is required when 'checksum_value' is specified")
}
```

A test for this case should also be added.

5.  `$checksum_value` Type Too Loose

The parameter is typed as `Optional[String]`, which permits an empty string `""`. For consistency with the rest of the module (e.g. `$content`, `$source`) and to catch user errors early, this should be:

```puppet
Optional[String[1]] $checksum_value = undef,
```

6. It would be preferred to create a separate PR for the document and alignment changes unrelated to the feature as it's difficult to review.

NeatNerdPrime commented:
nevermind this PR , closing
